src/mlpack/tests/string_encoding_test.cpp

```diff
@@ -0,0 +1,640 @@
+
+  vector<vector<size_t>> output;
+  BagOfWordsEncoding<SplitByAnyOf::TokenType> encoder(
+      (BagOfWordsEncodingPolicy()));
+  SplitByAnyOf tokenizer(" ");
+
+  encoder.Encode(stringEncodingInputSmall, output, tokenizer);
+
+  const DictionaryType& dictionary = encoder.Dictionary();
+
+  // Checking that everything is mapped to different numbers
+  std::unordered_map<size_t, size_t> keysCount;
+  for (auto& keyValue : dictionary.Mapping())
+  {
+    keysCount[keyValue.second]++;
+    // Every token should be mapped only once
+    BOOST_REQUIRE_EQUAL(keysCount[keyValue.second], 1);
+  }
+
+  vector<vector<size_t>> expected = {
+    { 1, 1, 1, 1, 0, 0, 0, 0 },
+    { 0, 0, 0, 0, 1, 1, 1, 0 },
+    { 0, 1, 1, 1, 0, 0, 0, 1 }
+  };
+
+  BOOST_REQUIRE(output == expected);
+}
+
+/**
+* Test Bag of Words encoding for characters using lamda function.
+*/
+BOOST_AUTO_TEST_CASE(BagOfWordsEncodingIndividualCharactersTest)
+{
+  vector<string> input = {
+    "GACCA",
+    "ABCABCD",
+    "GAB"
+  };
+
+  arma::mat output;
+  BagOfWordsEncoding<CharExtract::TokenType> encoder;
+
+  // Passing a empty string to encode characters
+  encoder.Encode(input, output, CharExtract());
+  arma::mat target = {
+    { 1, 1, 1, 0, 0 },
+    { 0, 1, 1, 1, 1 },
+    { 1, 1, 0, 1, 0 }
+  };
+
+  CheckMatrices(output, target);
+}
+
+/**
+ * Test the one pass modification of the Bag of Words encoding algorithm
+ * in case of individual character encoding.
+ */
+BOOST_AUTO_TEST_CASE(OnePassBagOfWordsEncodingIndividualCharactersTest)
+{
+  std::vector<string> input = {
+    "GACCA",
+    "ABCABCD",
+    "GAB"
+  };
+
+  vector<vector<size_t>> output;
+  BagOfWordsEncoding<CharExtract::TokenType> encoder;
+
+  // Passing a empty string to encode characters
+  encoder.Encode(input, output, CharExtract());
+
+  vector<vector<size_t>> expected = {
+    { 1, 1, 1, 0, 0 },
+    { 0, 1, 1, 1, 1 },
+    { 1, 1, 0, 1, 0 }
+  };
+
+  BOOST_REQUIRE(output == expected);
+}
+
+BOOST_AUTO_TEST_CASE(TfIdfEncodingTest)
+{
+  using DictionaryType = StringEncodingDictionary<boost::string_view>;
+
+  arma::mat output;
+  TfIdfEncoding<SplitByAnyOf::TokenType> encoder;
+  SplitByAnyOf tokenizer(" ");
+
+  encoder.Encode(stringEncodingInputSmall, output, tokenizer);
+
+  const DictionaryType& dictionary = encoder.Dictionary();
+
+  // Checking that everything is mapped to different numbers
+  std::unordered_map<size_t, size_t> keysCount;
+  for (auto& keyValue : dictionary.Mapping())
+  {
+    keysCount[keyValue.second]++;
+    // Every token should be mapped only once
+    BOOST_REQUIRE_EQUAL(keysCount[keyValue.second], 1);
+  }
+  arma::mat expected = {
+    { 0.1192803136799, 0.0440228147639, 0.0440228147639, 0.0440228147639, 0, 0, 0, 0 },
+    { 0, 0, 0, 0, 0.1590404182399, 0.1590404182399, 0.1590404182399, 0 },
+    { 0, 0.0440228147639 , 0.0440228147639 , 0.0440228147639 , 0, 0, 0, 0.1192803136799 }
+  };
+  CheckMatrices(output, expected, 1e-10);
```

Comment:
Why does your result differ from `scikit-learn`?

I tried the following test
```python
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer()

text = [
    "hello how are you",
    "i am good",
    "Good how are you"
]

X = vectorizer.fit_transform(text)

print(X)
```
The output was
```
  (0, 3)	0.6046521283053111
  (0, 4)	0.4598535287588349
  (0, 1)	0.4598535287588349
  (0, 5)	0.4598535287588349
  (1, 0)	0.7959605415681652
  (1, 2)	0.6053485081062916
  (2, 4)	0.5
  (2, 1)	0.5
  (2, 5)	0.5
  (2, 2)	0.5
```

Reply:
> Why does your result differ from `scikit-learn`?
> 
> I tried the following test
> 
> ```python
> from sklearn.feature_extraction.text import TfidfVectorizer
> 
> vectorizer = TfidfVectorizer()
> 
> text = [
>     "hello how are you",
>     "i am good",
>     "Good how are you"
> ]
> 
> X = vectorizer.fit_transform(text)
> 
> print(X)
> ```
> 
> The output was
> 
> ```
>   (0, 3)	0.6046521283053111
>   (0, 4)	0.4598535287588349
>   (0, 1)	0.4598535287588349
>   (0, 5)	0.4598535287588349
>   (1, 0)	0.7959605415681652
>   (1, 2)	0.6053485081062916
>   (2, 4)	0.5
>   (2, 1)	0.5
>   (2, 5)	0.5
>   (2, 2)	0.5
> ```

Ok This was very strange enough , I though Sklearn implemented the basic version of tfidf
but it is not so, They are using some other variant

https://en.wikipedia.org/wiki/Tf%E2%80%93idf#Term_frequency_2

`tfidf score = tf_value * idf_value`
idf_value doesn't have different variants

But tf_value has 4 types of variants, namely 

-     Boolean "frequencies": tf(t,d) = 1 if t occurs in d and 0 otherwise;
 

-     term frequency adjusted for document length : ft,d ÷ (number of words in d)

   

-     logarithmically scaled frequency: tf(t,d) = log (1 + ft,d);[6]

    

-     augmented frequency, to prevent a bias towards longer documents, e.g. raw frequency divided by the raw frequency of the most occurring term in the document:

I have implemented the second variant and thus there is difference in calculation 

Also I have noticed , there are different base of for idf values 
some implementation uses base 2, some used base e and some used base 10, but the Wikipedia page and standard implementation which most of them use is base 10

https://en.wikipedia.org/wiki/Tf–idf#Example_of_tf–idf

Reply:
I think we should implement the same algorithm as in `scikit-learn` by default since it's on of the most known packages. Probably it's a good idea to implement other techniques as well. It's possible since we can pass arguments to the policy constructor.

Reply:
Try the following formulas: [https://scikit-learn.org/stable/modules/feature_extraction.html#tfidf-term-weighting](https://scikit-learn.org/stable/modules/feature_extraction.html#tfidf-term-weighting).

Reply:
I am not sure, we will be able to provide support for normalization since The output matrix is created index by index , thus I guess we cannot support normalisation

Reply:
It's possible. We can add a policy function and call it at the end of each line. But I am not sure that it's needed. Probably it's better to write a number of scaler classes than transform the encoder to a huge combine. Let's do without the normalization.

Reply:
> Try the following formulas: https://scikit-learn.org/stable/modules/feature_extraction.html#tfidf-term-weighting.

Ok so I have added
Following variation 
  bool binary;
  bool sublinear_tf;
  bool raw_count;
  bool term_frequency;

and each of these can be coupled with   bool smooth_idf;
so we 8 different types 
Do i have to write test for all the 8 ? for both arma::mat and vector<vector<double>>
and also do i have to write it for both charextract and stringInput 

If we are taking all of these then we have 4*2*2*2 (24 test alone for tf-idf)